# -*- coding: utf-8 -*-
"""Cópia de script_24_10_25

Automatically generated by Colab.

Original file is located at
    https://colab.research.google.com/drive/1I2M0x1EmvFEiA65UZgPhk1FbFHKt2fKG
"""

# Dogma

#from google.colab import drive
#drive.mount('/content/drive')

!pip install Biopython

from Bio.Seq import Seq
from Bio.SeqRecord import SeqRecord
from Bio import SeqIO

from google.colab import drive
drive.mount('/content/drive')

Path = input("input file's path")
with open(Path,'r') as dna_file:
    gene = Seq(dna_file.read())

import re

# Transcrição o gene para um RNA mensageiro

def DNA_to_RNAm(dna_sequence):
  RNAm = ""
  for base in dna_sequence:
    if base == "A":
      RNAm += "U" # A no DNA pareia com U no RNA
    elif base == "T":
      RNAm += "A"
    elif base == "G":
      RNAm += "C"
    elif base == "C":
      RNAm += "G"
  return RNAm

mensageiro = DNA_to_RNAm(gene)
seq_mensageiro = Seq(mensageiro)

# Dicionário de condons

codon_amino_acid_map = {
    "UUU": "F", "UUC": "F", "UUA": "L", "UUG": "L",
    "CUU": "L", "CUC": "L", "CUA": "L", "CUG": "L",
    "AUU": "I", "AUC": "I", "AUA": "I", "AUG": "M",
    "GUU": "V", "GUC": "V", "GUA": "V", "GUG": "V",

    "UCU": "S", "UCC": "S", "UCA": "S", "UCG": "S",
    "CCU": "P", "CCC": "P", "CCA": "P", "CCG": "P",
    "ACU": "T", "ACC": "T", "ACA": "T", "ACG": "T",
    "GCU": "A", "GCC": "A", "GCA": "A", "GCG": "A",

    "UAU": "Y", "UAC": "Y", "UAA": "*", "UAG": "*",
    "CAU": "H", "CAC": "H", "CAA": "Q", "CAG": "Q",
    "AAU": "N", "AAC": "N", "AAA": "K", "AAG": "K",
    "GAU": "D", "GAC": "D", "GAA": "E", "GAG": "E",

    "UGU": "C", "UGC": "C", "UGA": "*", "UGG": "W",
    "CGU": "R", "CGC": "R", "CGA": "R", "CGG": "R",
    "AGU": "S", "AGC": "S", "AGA": "R", "AGG": "R",
    "GGU": "G", "GGC": "G", "GGA": "G", "GGG": "G"
}

# Translation

#Loop that scans the genetic and translates into aminoacids
protein = ""
for i in range(0, len(mensageiro),3):
  codon = mensageiro[i:i+3]
  aminoacid = codon_amino_acid_map.get(codon,"000")
  protein += aminoacid

protein_sequence = SeqRecord(
        Seq(protein),
        id="seq1",
        description="Protein sequence of gene"
    )

# The file is saved as 'protein_sequence.fasta' in the Colab environment.
SeqIO.write(protein_sequence, "protein_sequence.fasta", "fasta")

# To open and read the created FASTA file:
with open("protein_sequence.fasta", "r") as f_in:
    read_record = SeqIO.read(f_in, "fasta")
    print(f"ID: {read_record.id}")
    print(f"Description: {read_record.description}")
    print(f"Sequence: {read_record.seq}")
